protocol/strict-derivation.md

```diff
@@ -0,0 +1,269 @@
+is quite unlikely, unless the batcher implementation would be specially tuned to backfill gaps. If a
+gap occurs, it is actually much more likely that the next batches will be even further in the
+future, so option 2 would realistically almost always be a bad choice. Plus, the sync start
+algorithm benefits from option 1 as well, as it leads to faster resolution of uncertainty caused
+by gaps.
+
+It should be noted that if the blocks after the gap actually come from a chain where the gaps
+aren't deposit-only blocks, then these blocks would have mismatched parent hashes and be likely
+discarded. This is intentional. This design allows for not having to submit empty batches to
+derive deposit-only blocks, but instead just pick the next non-deposit-only block.
+
+Writing this, I realize that this mechanism could even be used to encode large spans of empty
+batches, as long as the sequencer is creating unsafe blocks that follow the same L1 origins as the
+auto-derivation would for gaps. However, this would need to be investigated more deeply.
+
+Note that the new strict ordering rules of the batch queue will always lead to an empty batch queue
+when the origin of the derivation pipeline progresses to the next L1 block (what about
+`undecided` though?).
+
+## Batcher hardening
+
+In a sense, Holocene shifts some complexity from derivation to the batching phase. Simpler but
+stricter derivation rules need to be met by a more complex batcher implementation.
+
+The batcher must be hardened to guarantee the strict ordering requirements. They are already mostly
+met in practice by the current implementation, but more by accident than by design. There are edge
+cases in which the batcher might violate the strict ordering rules. For example, if a channel fails
+to submit within a set period, the blocks are requeued and some out of order batching might occur.
+We also need to take extra care that dynamic blobs/calldata switching doesn't lead to out of order
+or gapped batches in scenarios where blocks are requeued, while future channels are already waiting
+in the mempool for inclusion.
+
+I propose to let the batcher follow a fixed nonce to block-range assignment, once the first batcher
+transaction (which is almost always the only batcher transaction for a channel) starts being
+submitted. This should avoid out-of-order or gapped batches. It might require to implement some form
+of persistence in the transaction manager, since we cannot reliably recover all globally pending
+batcher transactions in the L1 network.
+
+Furthermore, the batcher will need to be made aware of the Steady Batch Derivation rules, namely that
+invalid batches will be derived as deposit-only blocks. So in case of a reorg, the batcher should
+e.g. wait on the sequencer it is connected to until it has derived all blocks from L1 in order to
+only start batching new blocks on top of the possibly derived deposit-only chain segment.
```

Comment:
Thinking through this a bit, I'm having trouble deciding if this is or isn't a problem.

If an L1 reorg occurs, it only affects the L2 if a batch is wiped out.

If a batch is wiped out, then the batcher's nonce is reverted. Per Seb's proposal "a fixed nonce to block-range assignment", the batcher wants to resubmit on the same range as before, so there isn't a risk of the batcher posting a batch-gap (which would create empty blocks).

🤔 🤔 🤔

Reply:
This is a valid concern, and has been flagged by @tynes in his original design doc's [risk section](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/invalid-block-progression.md#risks--uncertainties).

The current rules of Steady Batch Derivation can indeed lead to a previously valid submitted batch to now be invalid, and then immediately be derived as deposit-only blocks, causing a long L2 unsafe reorg. The same batcher tx may still be included on L1. However, given that span batches are only forward-invalidated with Holocene, I think the L2 unsafe reorg would be limited to the L2 section that references the reorged-out L1 section. However, more batcher txs that might already have landed on L1 as well would cause more deposit-only blocks to be derived.

I think this is the tradeoff of Steady Batch Derivation.

One solution that I can think of that may alleviate this problem is to reference the last L1 origin in the channel metadata (in a new channel format), and then drop the channel directly in the channel bank before even decoding any batches from it, that would then at some point be derived as deposit-only blocks. This way, the batcher could get a "second chance" to submit a channel that includes the correct reorged-to L1 origin chain. This is very similar to how span batches contain the last L1 origin as `l1_origin_check` in their prefix, just moved one layer up to the channel container. With such a new channel format, the L1 origin check could arguably be dropped from span batches. Having the channel, rather than any span or singular batch, contain such a L1 origin check has the advantage that the DP wouldn't too eagerly derive deposit-only blocks, and to recover from L1 reorgs. I think this solution would also still maintain the nice properties of Strict Batch Ordering, that there's only one staging channel, and that we don't buffer out of order frames or batches.

If Proofs and Interop experts can confirm that such a solution would still lead to the sought after improvements for Proofs and Interop, resp., we could consider including it as part of Holocene.

Reply:
Added this to the design doc.

Reply:
Instead of doing this channel L1 origin check, actually just throw away span batches with failing L1 origin check, but give a second chance to be replaced, and don't generate deposit-only blocks for the whole span batch range.

Reply:
To fully benefit from this, only decode the prefix first, do the check.